dace/sdfg/analysis/cfg.py

```diff
@@ -59,6 +59,174 @@ def all_dominators(
+    # ipostdoms
+    g = nx.DiGraph()
+    for node, pdom in ipostdom.items():
+        if node is pdom:
+            continue
+        g.add_edge(node, pdom)
+    tc = nx.transitive_closure_dag(g)
+    all_postdoms: Dict[ControlFlowBlock, Set[ControlFlowBlock]] = defaultdict(set)
+    for node in tc:
+        all_postdoms[node] = set(dst for _, dst in tc.out_edges(node))
+
+    if remove_sink:
+        cfg.remove_node(sink)
+
+    return all_postdoms
+
+
+def find_sese_region(
+    graph: ControlFlowRegion, target_nodes: Set[ControlFlowBlock]
+) -> Tuple[Set[ControlFlowBlock], Optional[ControlFlowBlock], Optional[ControlFlowBlock]]:
+    """
+    Find the smallest SESE region containing the target nodes.
+
+    :param graph: The control flow graph to analyze.
+    :param target_nodes: The set of target nodes to include in the SESE region.
+    :param start_node: The starting node of the SESE region. If None, the start node of the graph is used.
+    :param end_nodes: The end node of the SESE region. If None, a virtual sink node is created temporarily.
+    :return: A tuple containing:
+        - A set of nodes in the SESE region.
+        - The entry node of the SESE region.
+        - The exit node of the SESE region.
+    :raises ValueError: If no start node or end nodes are found and none are provided.
+    """
+    if not target_nodes:
+        return set(), None, None
+
+    sinks = set()
+    for block in graph.nodes():
+        if graph.out_degree(block) == 0 or isinstance(block, (ContinueBlock, BreakBlock, ReturnBlock)):
+            sinks.add(block)
+    sink = ControlFlowBlock('__DACE_dummy_sink')
+    graph.add_node(sink)
+    for s in sinks:
+        graph.add_edge(s, sink, InterstateEdge())
+
+    # Compute dominators and post-dominators
+    dominators = all_dominators(graph)
+    post_dominators = all_postdominators(graph, sink=sink)
+
+    # Find the entry node: the lowest common dominator of all target nodes
+    common_dominators = None
+    for node in target_nodes:
+        if node not in dominators:
+            continue
+        if common_dominators is None:
+            common_dominators = dominators[node].copy()
+        else:
+            common_dominators &= dominators[node]
+
+    if not common_dominators:
+        return set(), None, None
+
+    # The entry is the dominator closest to the target nodes
+    entry_node = None
+    min_distance = float('inf')
+    for dom in common_dominators:
+        # Find maximum distance to any target node
+        max_dist_to_targets = 0
+        for target in target_nodes:
+            if target in dominators and dom in dominators[target]:
+                # Count nodes between dom and target
+                path_exists = nx.has_path(graph.nx, dom, target)
+                if path_exists:
+                    try:
+                        dist = nx.shortest_path_length(graph.nx, dom, target)
```

Comment:
Isn't running shortest_path_length enough? if there is no path it will raise the exception.

Reply:
You are right, changed it